### PR TITLE
Split binary and library to separate crates and make RPC feature

### LINKED ISSUE #298: Put RPC Storage under a dedicated feature
Add an `rpc-storage` feature that enables this module:
https://github.com/risechain/pevm/blob/7c9529e0f5435776c5f25af384d955e6dfdefb50/src/storage/rpc.rs#L1-L2

And these dependencies:
https://github.com/risechain/pevm/blob/7c9529e0f5435776c5f25af384d955e6dfdefb50/Cargo.toml#L43-L50

### PR DESCRIPTION
**Pull request by i1i1**
Closes #298

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -1,3 +1,3 @@
 [submodule "tests/ethereum/tests"]
-	path = tests/ethereum/tests
+	path = pevm/tests/ethereum/tests
```
Comment (hai-rise) [inline .gitmodules:2]: We also need to update the `working-directory` for lint and tests in CI, which would actually fail currently.

File: Cargo.toml [hunk 2]
```diff
@@ -1,89 +1,62 @@
-[package]
```
Comment (hai-rise) [inline Cargo.toml:1]: Nit but let's format TOML files as well for consistency.

File: Cargo.toml [hunk 3]
```diff
@@ -1,89 +1,62 @@
...
-[features]
-optimism = [
-    "revm/optimism",
-    "dep:op-alloy-consensus",
-    "dep:op-alloy-network",
-    "dep:op-alloy-rpc-types",
+[workspace]
+default-members = ["bins/*"]
+members = [
+  "bins/*",
```
Comment (hai-rise) [inline Cargo.toml:4]: Nit but I prefer to list all members instead of listing patterns.

File: Cargo.toml [hunk 4]
```diff
@@ -1,89 +1,62 @@
...
-edition = "2021"
-
-[features]
-optimism = [
-    "revm/optimism",
-    "dep:op-alloy-consensus",
-    "dep:op-alloy-network",
-    "dep:op-alloy-rpc-types",
+[workspace]
+default-members = ["bins/*"]
```
Comment (hai-rise) [inline Cargo.toml:2]: This should be the core `pevm` library?

File: bins/fetch/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,19 @@
+[package]
+name = "pevm-fetch"
+version = "0.1.0"
+edition = "2021"
+
+[dependencies]
+bincode.workspace = true
+clap = { workspace = true, features = ["derive"] }
+flate2.workspace = true
+serde_json = { workspace = true, features = ["preserve_order"] }
```
Comment (hai-rise) [inline bins/fetch/Cargo.toml:10]: Good features all crates enable by default can be better declared in the workspace `Cargo.toml`?

File: Cargo.toml [hunk 6]
```diff
@@ -1,89 +1,62 @@
...
-[[bench]]
-name = "gigagas"
-harness = false
-
-[profile.release]
-codegen-units = 1
-panic = "abort"
-lto = "fat"
+tokio = "1.40.0"
+walkdir = "2.5.0"
```
Comment (hai-rise) [inline Cargo.toml:62]: Nit but let's add final trailing newlines :pray:.

File: pevm/Cargo.toml [hunk 7]
```diff
@@ -0,0 +1,80 @@
+[package]
+name = "pevm"
+version = "0.1.0"
+edition = "2021"
+
+[features]
+defaults = []
+rpc = [
+  "tokio",
```
Comment (hai-rise) [inline pevm/Cargo.toml:9]: These should be `dep:*`? Nit but let's also sort the deps alphabetically and add a newline to separate each feature set.

File: pevm/src/lib.rs [hunk 8]
```diff
@@ -221,9 +221,11 @@ mod pevm;
 pub use pevm::{execute_revm_sequential, Pevm, PevmError, PevmResult};
 mod scheduler;
 mod storage;
+#[cfg(feature = "rpc")]
```
Comment (hai-rise) [inline pevm/src/lib.rs:224]: Nit but I'd put it below the default list.

**Review by hai-rise — CHANGES_REQUESTED** [review]
I have left a few nits. We also need to make the CI lint & tests `--workspace` and fix accordingly (it's failing).

`pevm/pevm` is a bit uncomfortable but overall I like this direction. One day we may have another library crate to make it `pevm/crates/{pevm, ???}`, too :).

File: bins/fetch/Cargo.toml [hunk 9]
```diff
@@ -0,0 +1,19 @@
+[package]
+name = "pevm-fetch"
```
Comment (hai-rise) [inline bins/fetch/Cargo.toml:2]: I like this name better. Let's also update it in the README.

File: pevm/Cargo.toml [hunk 10]
```diff
@@ -0,0 +1,80 @@
+[package]
```
Comment (hai-rise) [inline pevm/Cargo.toml:1]: Let's add a `crates` directory to make it more "standard", so this should be `pevm/crates/pevm/Cargo.toml`.

File: Cargo.toml (hunk 2 above)
Comment (i1i1) [reply Cargo.toml:1]: Should I use https://github.com/tamasfe/taplo or smth else for formatting?

File: Cargo.toml (hunk 4 above)
Comment (i1i1) [reply Cargo.toml:2]: Hmm, this just means that when we'll do `cargo build` it'll build binaries, but I can change it

File: bins/fetch/Cargo.toml (hunk 5 above)
Comment (i1i1) [reply bins/fetch/Cargo.toml:10]: You mean, that I should enable some features on workspace level, right? Should I add smth else, apart from `serde_json`?

File: Cargo.toml (hunk 2 above)
Comment (hai-rise) [reply Cargo.toml:1]: Yes, that is a good choice. I'll add it to the upcoming `CONTRIBUTING.md`!

File: Cargo.toml (hunk 4 above)
Comment (hai-rise) [reply Cargo.toml:2]: Yeah, the binary is very niche so most of time we lint, build, test, and bench the library by default (for development).

File: bins/fetch/Cargo.toml (hunk 5 above)
Comment (hai-rise) [reply bins/fetch/Cargo.toml:10]: Yes, let's port all feature choices to the workspace `Cargo.toml` for simplicity.

File: .github/workflows/ci.yml [hunk 11]
```diff
@@ -25,7 +25,7 @@ jobs:
           key: ${{ runner.os }}-cargo-${{ hashFiles('**/Cargo.lock') }}
 
       - name: Lint
-        run: cargo clippy --all-targets --all-features
+        run: cargo clippy --all --all-targets --all-features
```
Comment (hai-rise) [inline .github/workflows/ci.yml:28]: Let's use `--workspace`:
```
Package Selection:
  -p, --package [<SPEC>]  Package to build (see `cargo help pkgid`)
      --workspace         Build all packages in the workspace
      --exclude <SPEC>    Exclude packages from the build
      --all               Alias for --workspace (deprecated)
```

File: README.md [hunk 12]
```diff
@@ -77,10 +77,10 @@ $ cargo build
 
 #### Fetch Real Blocks
```
Comment (hai-rise) [inline README.md:78]: Can you also update the paths in `Testing` and `Benchmarks`?

File: bins/fetch/Cargo.toml (hunk 5 above)
Comment (hai-rise) [reply bins/fetch/Cargo.toml:10]: You've only done it for `pevm`, let's do the same here.

File: crates/pevm/Cargo.toml [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+]
+
+[dependencies]
+alloy-chains.workspace = true
+alloy-consensus.workspace = true
+alloy-primitives.workspace = true
+alloy-rlp.workspace = true
+alloy-rpc-types.workspace = true
+alloy-trie.workspace = true
+# We can roll our own but [revm] depends on this anyway.
```
Comment (hai-rise) [inline crates/pevm/Cargo.toml:28]: Very nit but this comment is better ported to the workspace `Cargo.toml` now.

File: crates/pevm/tests/common/mod.rs [hunk 14]
```diff
@@ -54,36 +54,35 @@ pub const RAW_TRANSFER_GAS_LIMIT: u64 = 21_000;
...
+        File::open(data_dir.join("bytecodes.bincode.gz")).unwrap(),
     )))
     .unwrap();
 
-    for block_path in fs::read_dir("data/blocks").unwrap() {
+    for block_path in fs::read_dir(data_dir.join("blocks")).unwrap() {
         let block_path = block_path.unwrap().path();
         let block_number = block_path.file_name().unwrap().to_str().unwrap();
 
+        let dir = data_dir.join("blocks").join(block_number);
```
Comment (hai-rise) [inline crates/pevm/tests/common/mod.rs:69]: `block_dir` for clarity :pray:.

**Review by hai-rise — CHANGES_REQUESTED** [review]
Almost there!

File: crates/pevm/Cargo.toml [hunk 15]
```diff
@@ -0,0 +1,76 @@
+[package]
+name = "pevm"
+version = "0.1.0"
+edition = "2021"
+
+[features]
+defaults = []
+rpc = [
```
Comment (hai-rise) [inline crates/pevm/Cargo.toml:8]: Oh one last thing but it should be `rpc-storage` since `rpc` is very broad and can be misleading.

**Review by hai-rise — APPROVED** [review]
Really like the new structure, thanks! Also appreciated that you suggested further improvements over what was needed 🙏.

**Pull request merged**